Extended resolve default cache to improve the resiliency and protect from DNS infrastructure failure and outages

KiranVKumar opened the pull request:
Introduces a new feature, the `StaleRetentionSec` flag, to the systemd-resolved service. This flag allows users to enhance DNS resiliency by utilizing the serve stale feature, which safeguards client applications from DNS server failures. This feature is an implementation inspired by the DNS Serving Stale Data mechanism outlined in [RFC 8767](https://datatracker.ietf.org/doc/rfc8767/). 

When the `StaleRetentionSec `property is set, the systemd-resolved service retains DNS resource records in the cache longer than their standard Time to Live (TTL), causing these records to be classified as stale data.  systemd-resolved responds with the stale data in case of DNS server failures or timeouts (except NXDOMAIN) to protect client applications from DNS-related disruptions.

bluca commented:
```
../src/resolve/resolved-dns-transaction.c:1712:53: error: address of 't->scope->manager->enable_serve_stale' will always evaluate to 'true' [-Werror,-Wpointer-bool-conversion]
                                &t->scope->manager->enable_serve_stale && t->n_attempts > 1);
                                 ~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~ ~~
```

also indentation is off in many places, please fix that

ddstreet commented:
This would explicitly violate the dns rfc which says caching servers should not use any data past its TTL. Can you explain why this is needed (what problem does this fix)?

KiranVKumar commented:
> This would explicitly violate the dns rfc which says caching servers should not use any data past its TTL. Can you explain why this is needed (what problem does this fix)?

This will provide protection to the client application against DNS infrastructure failures and outages. This is similar to RFC 8767. We have an open feature request https://github.com/systemd/systemd/issues/21815 someone has made a request for a similar solution.

poettering commented:
> This would explicitly violate the dns rfc which says caching servers should not use any data past its TTL. Can you explain why this is needed (what problem does this fix)?

I think the PR makes sense. If you have fundamental problems contacting your DNS server then stale data (as determined by TTL) is better than no data I am sure.

poettering commented on src/resolve/resolved-dns-cache.c:
```diff
@@ -54,6 +54,7 @@ struct DnsCacheItem {
         LIST_FIELDS(DnsCacheItem, by_key);
 
         bool shared_owner;
+        usec_t ttl_expiry;        /* The key is for storing the time when the TTL set to expire. */
```
given this is closely related to "until", I think it should be defined close to it

KiranVKumar replied:
renamed to until_valid

poettering commented on src/resolve/resolved-dns-cache.c:
```diff
@@ -411,7 +414,8 @@ static int dns_cache_put_positive(
                 int owner_family,
-                const union in_addr_union *owner_address) {
+                const union in_addr_union *owner_address,
+                bool enable_serve_statle) {
```
typo

KiranVKumar replied:
Fixed

poettering commented on src/resolve/resolved-gperf.gperf:
```diff
@@ -32,3 +32,4 @@ Resolve.ReadEtcHosts,              config_parse_bool,                    0,
 Resolve.ResolveUnicastSingleLabel, config_parse_bool,                    0,                   offsetof(Manager, resolve_unicast_single_label)
 Resolve.DNSStubListenerExtra,      config_parse_dns_stub_listener_extra, 0,                   offsetof(Manager, dns_extra_stub_listeners)
 Resolve.CacheFromLocalhost,        config_parse_bool,                    0,                   offsetof(Manager, cache_from_localhost)
+Resolve.ServeStale,                config_parse_bool,                    0,                   offsetof(Manager, enable_serve_stale)
```
I think this should take a time vlaue, and specify the hard TTL when we'll stop serving stale data. And if people set this to "0" then we will disable the feature, and if they set it to "infinity" then we'll keep the stale data forever.

config_parse_usec already parses "infinity" for you, hence this shouldn't be too hard.

poettering commented on src/resolve/resolved-dns-cache.c:
```diff
@@ -477,13 +482,17 @@ static int dns_cache_put_positive(
                 .answer = dns_answer_ref(answer),
                 .full_packet = dns_packet_ref(full_packet),
-                .until = calculate_until(rr, min_ttl, UINT32_MAX, timestamp, false),
+                .until = enable_serve_statle ? timestamp + USEC_PER_DAY : ttl_expiry,
```
usec_add() to deal with overflows somewhat sensibly

poettering commented on src/resolve/resolved-dns-cache.c:
```diff
@@ -54,6 +54,7 @@ struct DnsCacheItem {
 
         bool shared_owner;
+        usec_t ttl_expiry;        /* The key is for storing the time when the TTL set to expire. */
 };
```
i'd prefer if we'd store two time-values `until_valid`, `until_stale` or so, where the first one is the time the entry remains fully valid by the indicated TTL and the second one indicates the time it remains "vaguely valid" as per the stale rules.

I think that would simplify things a lot, as the two timeouts are then similarly named, and behave similarly, we just use either one or the other timeout depending on context.

KiranVKumar replied:
Renamed `ttl_expiry `to `until_valid`.

poettering commented on src/resolve/resolved-dns-cache.c:
```diff
@@ -533,6 +543,10 @@ static int dns_cache_put_negative(
         assert(owner_address);
 
+        /* Don't store the negative responses if the 'ServeStale' flag is enabled. */
+        if(enable_serve_statle && rcode != DNS_RCODE_NXDOMAIN)
```
follow coding style, i.e. a space before the `(`

KiranVKumar replied:
Fixed

poettering commented on src/resolve/resolved-dns-cache.c:
```diff
@@ -1024,6 +1043,15 @@ int dns_cache_lookup(
                 }
 
+                /* Skip the next part if ttl is expired in first attempt*/
+                if( !use_stale_data && j->ttl_expiry < current){
```
coding style, space *before* the "(", not after.

and space between ) and {

KiranVKumar replied:
Fixed

poettering commented:
also needs man page update to document the new setting

ddstreet commented on src/resolve/resolved-manager.c:
```diff
@@ -600,6 +600,7 @@ int manager_new(Manager **ret) {
 
                 .sigrtmin18_info.memory_pressure_handler = manager_memory_pressure,
                 .sigrtmin18_info.memory_pressure_userdata = m,
+                .enable_serve_stale = false,
```
> hmm, i actually would prefer if this feature would be enabled by default. I genuinely believe that stale info is always better than total failure. maybe default to something mildly conservative such as 6h though

I would advise against making this enabled by default; RFC 1035 clearly states that RR with expired TTL "should" be ignored. I believe @KiranVKumar is implementing RFC 8767 here, and you can see the security implications as outlined by the rfc: https://datatracker.ietf.org/doc/html/rfc8767#name-security-considerations

poettering replied:
tbh i don't find those convincing. I'd still turn it on by default. but i also don't care too much, up to you. would merge it either way

ddstreet commented:
If this is intending to implement RFC 8767, I think it should follow that RFC more closely? I've only given it a quick scan, but I think this PR is missing some things, like continuing to not cache any RR with a TTL of 0. I think this PR should both explicitly clarify (in code comments) that the 'serve_stale' option is an implementation of RFC 8767, as well as make sure the implementation adheres to that RFC.

ddstreet commented:
After reading rfc8767, I think these are the main points:

1) This RFC overrides RFC2181, and requires the TTL to be interpreted as unsigned int, instead of signed int. Meaning the current code would need to stop converting "negative" (i.e. MSB set) ttl values to 0. (IMHO this RFC's reasoning seems right for all cases on this point).
2) While our current code caps TTLs at 2 hours (which IMHO seems FAR to short), this RFC recommends 7 days. Personally, I think 7 days (max TTL) is much more sensible than 2 hours, but in any case, *IFF* systemd-resolved is serving "stale" DNS data, it makes absolutely no sense to cap TTL values at 2 hours but then continue to serve "stale" data (which isn't actually stale).
3) The RFC recommends the "stale timer" be user-configurable.
4) The RFC states that any "stale" data provided MUST include a TTL greater than 0, but it recommends all "stale" DNS data use a TTL of 30 seconds. Personally I think this is very appropriate; any "stale" DNS data should not be cached (by clients) for long - meaning, we should not be telling clients that stale DNS data has long TTL values.
5) Any upstream (authoritative) response of NoError or NXDOMAIN "must" clear/replace any existing cached data. I think this is our code's current behavior; just listing here to be sure it's confirmed.

Also, note that the RFC makes clear that after a DNS TTL has expired, the resolver MUST attempt to get fresh data from its upstream nameserver; it CANNOT continue to serve stale data without first attempting to actually resolve the name first. And this really is critical; if systemd-resolved caches DNS entries long beyond their actual TTL *without* attempting to refresh stale DNS data, it will absolutely break users.

ddstreet commented on src/resolve/resolved-dns-cache.c:
```diff
@@ -533,6 +546,10 @@ static int dns_cache_put_negative(
+        /* Don't store the negative responses if the StaleRetentionSec is greater than zero. */
+        if (stale_retention_usec > 0 && rcode != DNS_RCODE_NXDOMAIN)
+                        return 0;
+
```
No, don't do this. Below, the rcode is filtered down to either success, nxdomain, or servfail. The RFC explicitly states either success or nxdomain *must* replace our cached entry. So only when the rcode is servfail, and serve_stale is enabled, should we return 0. I think you should update the if check below.

ddstreet commented:
Looking over the current changes, I think this will be simpler and easier to manage if a separate 'expired' cache is used, instead of overloading ->until and adding ->until_valid. However, @poettering may disagree.

I also think this needs to adhere to the rfc, specifically the points in my [earlier comment](https://github.com/systemd/systemd/pull/27681#issuecomment-1562069484).

KiranVKumar commented:
> Looking over the current changes, I think this will be simpler and easier to manage if a separate 'expired' cache is used, instead of overloading ->until and adding ->until_valid. However, @poettering may disagree.
> 
> I also think this needs to adhere to the rfc, specifically the points in my [earlier comment](https://github.com/systemd/systemd/pull/27681#issuecomment-1562069484).

@ddstreet  

-  Modified the server's stale flag from a Boolean to a timer, allowing users to customize the duration for retaining stale data.
- The default caching behavior remains unchanged, where data is not cached if the time-to-live (TTL) is less than zero.
- Additionally, if a query is answered using stale data, the TTL will be set to the minimum value between the record's TTL and 30 seconds.
- The 'Serve Stale' feature is designed to protect client applications from server failures such as timeouts or ServFail responses. But if we let negative responses  clear the cache, the whole point of the 'Serve Stale' thing would be lost. That's why I added the condition that if the 'Serve Stale' flag is on, only NXDomain or positive responses can clear the cache.

KiranVKumar commented:
> also needs man page update to document the new setting

Updated resolve.config.xml with new property details

evverx commented:
I think the commit message is too vague. It should include links to the RFC (along with the reasons why resolved deviates from it in some places) and the issues on GitHub this PR is supposed to address.

There are no tests as far as I can see.

KiranVKumar commented:
> also needs man page update to document the new setting

Updated man page with new configuration details

poettering commented:
please squash your commits so that each commit is a logical step that compiles cleanly and stands on its own feet, and not necessarily a historical step.

We care a lot about being able to bisect cleanly, hence every single commit should make sense on its own, come with a proper commit msg, and compile cleanly.

poettering commented on man/resolved.conf.xml:
```diff
@@ -335,6 +335,13 @@ DNSStubListenerExtra=udp:[2001:db8:0:f102::13]:9953</programlisting>
+        <listitem><para>Takes a duration value to determine for how long DNS resource records can be retained in the cache beyond their Time To Live (TTL) to return them as stale records.
+        This is useful when a DNS server failure occurs or not reachable, as systemd-resolved will continue to use these stale records to answer the DNS questions in case of a negative response from the upstream servers.
+        This improves the resilience against DNS infrastructure failures and outages.
+        However, this doesn't apply to NXDOMAIN responses, as those are still perfectly valid responses.</para></listitem>
```
should mention the default value if not specified, and make clean what setting this to 0 means, and what setting this to infinity means

poettering replied:
also, please add an appropriate line to src/resolve/resolved.conf.in, commented out with an `#` just like the other lines, with the default value. This is useful for people who modify configuration as they can immediately see what the default is, and what the concept is called.

KiranVKumar replied:
Updated and #StaleRetentionSec=86400 to src/resolve/resolved.conf.in

poettering commented on man/resolved.conf.xml:
```diff
@@ -335,6 +335,13 @@ DNSStubListenerExtra=udp:[2001:db8:0:f102::13]:9953</programlisting>
+      <varlistentry>
+        <term>StaleRetentionSec=<replaceable>SECONDS</replaceable></term>
+        <listitem><para>Takes a duration value to determine for how long DNS resource records can be retained in the cache beyond their Time To Live (TTL) to return them as stale records.
+        This is useful when a DNS server failure occurs or not reachable, as systemd-resolved will continue to use these stale records to answer the DNS questions in case of a negative response from the upstream servers.
```
the way i see this a "negative response" is NXDOMAIN, but this is expressly not when we return the stale data, as your sentence further down then expressly makes clear.

Hence maybe say:

"… in case no valid answer can be acquired from upstream DNS servers."

KiranVKumar replied:
Updated

poettering commented on man/resolved.conf.xml:
```diff
@@ -335,6 +335,13 @@ DNSStubListenerExtra=udp:[2001:db8:0:f102::13]:9953</programlisting>
+        This is useful when a DNS server failure occurs or not reachable, as systemd-resolved will continue to use these stale records to answer the DNS questions in case of a negative response from the upstream servers.
+        This improves the resilience against DNS infrastructure failures and outages.
+        However, this doesn't apply to NXDOMAIN responses, as those are still perfectly valid responses.</para></listitem>
+      </varlistentry>
```
event though the above kinda says this already I think it would make sense to say expressly in a separate sentence that before resolved returns stale data it will always try to contact the upstream DNS servers first.

KiranVKumar replied:
Added

poettering commented on src/resolve/resolved-dns-cache.c:
```diff
@@ -18,6 +18,9 @@
 #define CACHE_TTL_MAX_USEC (2 * USEC_PER_HOUR)
 
+/* The max clamp tll for stale data is set to 30seconds. */
+#define CACHE_STALE_CLAMP_TTL_MAX_USEC (30 * USEC_PER_SEC)
```
why so short?

poettering replied:
or hmm, let me rephrase this: why so long?

i think we should probably set this to 1s, because we are already a local cache, and we'd rather have clients ask us again then to cache our response... given that talking to us, a local service is cheap effectively disabling caching in clients further up the pipeline should not be an issue

poettering commented on src/resolve/resolved-dns-cache.c:
```diff
@@ -54,6 +57,7 @@ struct DnsCacheItem {
         LIST_FIELDS(DnsCacheItem, by_key);
 
         bool shared_owner;
+        usec_t until_valid;        /* The key is for storing the time when the TTL set to expire. */
```
please move this up, next to where `until` is already defined.

poettering commented on src/resolve/resolved-dns-cache.c:
```diff
@@ -533,6 +546,10 @@ static int dns_cache_put_negative(
 
+        /* Don't store the negative responses if the StaleRetentionSec is greater than zero. */
+        if (stale_retention_usec > 0 && rcode != DNS_RCODE_NXDOMAIN)
+                        return 0;
```
indentation off

poettering commented on src/resolve/resolved-dns-cache.c:
```diff
@@ -533,6 +546,10 @@ static int dns_cache_put_negative(
         assert(owner_address);
 
+        /* Don't store the negative responses if the StaleRetentionSec is greater than zero. */
+        if (stale_retention_usec > 0 && rcode != DNS_RCODE_NXDOMAIN)
```
this looks wrong to me. this shouldn't be filtered when storing negative answers in the cache, but when using them from the cache. After all negative caching should always be used when TTL allows it, even if stale retention is on.

also DNS_RCODE_SUCCESS should be treated like DNS_RCODE_NXDOMAIN here, i.e. for NODATA cached entries.

KiranVKumar replied:
Removed the filter condition while putting the negative cache

poettering commented on src/resolve/resolved-dns-cache.c:
```diff
@@ -1068,15 +1099,19 @@ int dns_cache_lookup(
+                                 * If the question is resolved using stale data, the clamp TTL will be set as the minimum value between the TTL and 30 seconds. */
                                 DNS_ANSWER_FOREACH_ITEM(item, j->answer) {
-                                        r = answer_add_clamp_ttl(
+                                               r = answer_add_clamp_ttl(
```
hu?

poettering commented on src/resolve/resolved-dns-cache.c:
```diff
@@ -1091,7 +1126,7 @@ int dns_cache_lookup(
                                         NULL,
                                         query_flags,
-                                        j->until,
+                                        use_stale_data ? (current + MIN(j->rr->ttl * USEC_PER_SEC, CACHE_STALE_CLAMP_TTL_MAX_USEC)) : j->until_valid,
```
given you have the same expression basically twice here and above. can you wrap this in a tiny helper function?

poettering replied:
also, the ttl is a relative value, taken to the time the record was acquired, but you add it here to the current time, that isn't right!

poettering commented on src/resolve/resolved-dns-transaction.c:
```diff
@@ -1697,6 +1698,12 @@ static int dns_transaction_prepare(DnsTransaction *t, usec_t ts) {
+                 * On the second attempt, if StaleRetentionSec is greater than zero, try to answer the question using stale date (honors until property) */
+                bool use_stale_data = false;
+                if (t->scope->manager->stale_retention_usec > 0 && t->n_attempts > 1)
+                        use_stale_data = true;
```
```c
bool use_stale_data = t->scope->manager->stale_retention_usec > 0 && t->n_attempts > 1;
```

KiranVKumar commented on src/resolve/resolved-dns-cache.c:
```diff
@@ -42,7 +45,7 @@ struct DnsCacheItem {
         DnsPacket *full_packet;  /* The full packet this information was acquired with */
 
-        usec_t until;
+        usec_t until;            /*  If StaleRetentionSec is greater than zero, until is set to a duration of StaleRetentionSec from the date of creation. If StaleRetentionSec is zero, both until and ttl_expiry will set to ttl. */
```
updated

poettering commented on src/resolve/resolved-dns-cache.c:
```diff
@@ -1056,6 +1080,10 @@ int dns_cache_lookup(
 
+                /* If the question is being resolved using stale data, the clamp TTL will be set to one second. */
+                usec_t until = FLAGS_SET(query_flags, SD_RESOLVED_NO_STALE) ? j->until_valid
+                                                                            : (current + CACHE_STALE_CLAMP_TTL_MAX_USEC);
```
usec_add()

poettering commented on src/resolve/resolved-dns-transaction.c:
```diff
@@ -1697,10 +1698,16 @@ static int dns_transaction_prepare(DnsTransaction *t, usec_t ts) {
+                 * On the second attempt, if StaleRetentionSec is greater than zero, try to answer the question using stale date (honors until property) */
+                uint64_t query_flags = t->query_flags;
+                if (t->n_attempts == 1 || t->scope->manager->stale_retention_usec == 0)
+                        SET_FLAG(query_flags, SD_RESOLVED_NO_STALE, true);
```
just do:

```c
query_flags |= SD_RESOLVED_NO_STALE;
```

use the SET_FLAG() macro only if you actually have a boolean variable you want to propagate into a flag, but here you don't you just want to enable the flag

poettering commented on src/resolve/resolved.conf.in:
```diff
@@ -32,3 +32,4 @@
 #DNSStubListenerExtra=
 #ReadEtcHosts=yes
 #ResolveUnicastSingleLabel=no
+#StaleRetentionSec=86400
```
that said, the value shown here in the comments should be the implied default if the setting is not used. Hence, if we leave the option off by default, then this should be set to 0

KiranVKumar replied:
Changed to 1d as not enabling the config by default.

poettering commented:
The first commit adds some stuff the later commits then remove again (ttl_expiry). We generally don't like that. We want perfectly bisectable PRs, i.e. where every commit makes sense on its own, and where each commit is a logical step, but not a historical one. This makes working with the codebase and reviewing a lot easier..

Hence, please split up the PR differently in commits, so that each commit is a logical, additive step, but not a step that undoes something an earlier commit in the same PR just did.

(it might be fine to just squash all three commits into one, given the whole thing is not that complex)

`git rebase -i` is your friend

poettering commented:
codewise looks great, just some nitpicks. should be ready to be merged quickly

ddstreet reviewed (changes requested):
I still think the stale cache should be a separate cache (for simplicity in using it, for example it would be great if systemd-resolve could flush the stale cache independently, and if SIGUSR1 dumped the 'real' and 'stale' caches separately). However I think @poettering is ok with the approach here. Anyway, I added a few specific review comments to the current commits.

KiranVKumar commented on src/resolve/resolved-dns-cache.c:
```diff
@@ -1024,6 +1039,15 @@ int dns_cache_lookup(
+                if (FLAGS_SET(query_flags, SD_RESOLVED_NO_STALE) && j->until_valid < current) {
+                        log_debug("First attempt: TTL expired for %s Use stale data: %d",
+                                                dns_resource_key_to_string(key, key_str, sizeof key_str),
+                                                !FLAGS_SET(query_flags, SD_RESOLVED_NO_STALE));
```
removed the logging variable and renamed the message to "Requested with no stale and TTL expired for %s"

KiranVKumar commented on src/resolve/resolved-dns-transaction.c:
```diff
@@ -1716,6 +1723,14 @@ static int dns_transaction_prepare(DnsTransaction *t, usec_t ts) {
+                                /* log as info if the question is answered using stale data. */
+                                if (t->n_attempts > 1 && !FLAGS_SET(query_flags, SD_RESOLVED_NO_STALE)) {
+                                        char key_str[DNS_RESOURCE_KEY_STRING_MAX];
+                                        log_info("Serve Stale response rcode=%s for %s",
```
done

KiranVKumar commented on src/resolve/resolved-dns-cache.c:
```diff
@@ -18,6 +18,9 @@
 #define CACHE_TTL_MAX_USEC (2 * USEC_PER_HOUR)
 
+/* The max clamp tll for stale data is set to one second. */
+#define CACHE_STALE_CLAMP_TTL_MAX_USEC (1 * USEC_PER_SEC)
```
changed to 30sec as per the RFC 8767 section 6

KiranVKumar commented on src/resolve/resolved-dns-cache.c:
```diff
@@ -1024,6 +1039,15 @@ int dns_cache_lookup(
                         goto miss;
                 }
 
+                /* Skip the next part if ttl is expired in first attempt*/
```
updated comment.

KiranVKumar commented on src/resolve/resolved-dns-cache.c:
```diff
@@ -18,6 +18,9 @@
 #define CACHE_TTL_MAX_USEC (2 * USEC_PER_HOUR)
 
+/* The max clamp tll for stale data is set to one second. */
+#define CACHE_STALE_CLAMP_TTL_MAX_USEC (1 * USEC_PER_SEC)
```
Updated

KiranVKumar commented on src/resolve/resolved-dns-cache.c:
```diff
@@ -18,6 +18,9 @@
 /* We never keep any item longer than 2h in our cache */
 #define CACHE_TTL_MAX_USEC (2 * USEC_PER_HOUR)
```
CACHE_TTL_MAX_USEC  is used for calculating `until_valid `and it is default behavior , so keeping it as it.

ddstreet replied:
This still makes no sense to me. Not sure if @poettering agrees with me or not.

The only benefit I can see for clamping TTL is to reduce how long records stay in our cache. However, if we are keeping the record in our cache past the clamped TTL, there really wasn't any point in clamping it at all, in fact clamping the TTL actually makes everything *worse*.

poettering replied:
note that the clamping happens at cache lookup time, not at the time the cache is populated. that means what is actually stored in the cache is independent from the clamped TTL

KiranVKumar commented on src/resolve/resolved-dns-cache.c:
```diff
@@ -386,7 +391,8 @@ static void dns_cache_item_update_positive(
 
-        i->until = calculate_until(rr, min_ttl, UINT32_MAX, timestamp, false);
+        i->until_valid = calculate_until(rr, min_ttl, UINT32_MAX, timestamp, false);
+        i->until = stale_retention_usec > 0 ? usec_add(timestamp, stale_retention_usec) : i->until_valid;
```
updated it will be from the time of expiry.

KiranVKumar commented on src/resolve/resolved-dns-cache.c:
```diff
@@ -1056,6 +1080,10 @@ int dns_cache_lookup(
                         dnssec_result = _DNSSEC_RESULT_INVALID;
                 }
 
+                /* If the question is being resolved using stale data, the clamp TTL will be set to one second. */
```
updated

KiranVKumar commented on src/resolve/resolved-dns-cache.c:
```diff
@@ -386,7 +391,8 @@ static void dns_cache_item_update_positive(
         DNS_PACKET_REPLACE(i->full_packet, dns_packet_ref(full_packet));
 
-        i->until = calculate_until(rr, min_ttl, UINT32_MAX, timestamp, false);
+        i->until_valid = calculate_until(rr, min_ttl, UINT32_MAX, timestamp, false);
```
I have changed the `until` to set to a duration of `StaleRetentionSec` from the time of TTL expiry.

ddstreet commented on src/resolve/resolved.conf.in:
```diff
@@ -32,3 +32,4 @@
 #DNSStubListenerExtra=
 #ReadEtcHosts=yes
 #ResolveUnicastSingleLabel=no
+#StaleRetentionSec=1d
```
I think the default is 0, isn't it?

KiranVKumar replied:
Fixed. changed to 0

ddstreet commented on src/resolve/resolved-dns-cache.c:
```diff
@@ -18,6 +18,9 @@
 /* We never keep any item longer than 2h in our cache */
 #define CACHE_TTL_MAX_USEC (2 * USEC_PER_HOUR)
```
(I commented in a resolved thread but i'm not sure how to un-resolve it, so adding a new review)

This (clamping dns TTL values to 2 hours, but also keeping "stale" records after we think they "expire") still makes no sense to me. Not sure if @poettering agrees with me or not.

The only benefit I can see for clamping TTL is to reduce how long records stay in our cache. However, if we are keeping the record in our cache past the clamped TTL, there really wasn't any point in clamping it at all, in fact clamping the TTL actually makes everything worse.

poettering replied:
so my thinking was that overly long dns cache times don't really make sense on a client. we'd rather ask too often then too seldom. we are not a load-bearing DNS server for a bunch of clients, we are basically just an improvemet over the cache-less in-glibc resolver. Hence in the interest in being able to somewhat gracefully handle DNS misconfiguration i figured it makes sense to revalidate all entries after 2h no matter what, even if they are supposedly still valid for a year.

dunno. i tink this should be discussed separately from this PR.

poettering replied:
so i looked into this in more detail. I think the confusion comes from the fact how CACHE_TTL_MAX_USEC is applied. 

If the stale logic is enabled it means that if a request comes in later than 2h after caching we'll try to refresh every record. If it works, then great, we'll return the refreshed RRset. But if it doesn't, then we'll return the stale data, up to the configured stale rentation timeout. 

If the stale logic is disabled it means that if a request comes in later than 2h after caching we'll try to refresh it. If it works, great. If not, we'll return an error.

I think this is a reasonable approach I think. The question of course is what is more important to people: honour superlong ttls because it allows taping over connectivity problems (A), or limit superlong ttls because it allows taping over DNS configuration problems (B). I think when figuring out the answer to this one should take into consideration whether one is looking at a DNS resolver such as resolved that is on the end-user's system such as systemd-resolved or whether one is looking at a DNS resolver at an ISP or so. An ISP DNS resolver will have many upstreams a client DNS resolver pretty much only one it strictly relies on.

ddstreet reviewed (commented):
The config file example should show the actual default, not 1 day, and also I think @poettering should comment on whether to leave our TTL clamp at 2 hours or not.

It would be nice to have test(s) added as well for this functionality...

Otherwise LGTM.

poettering commented:
> I still think the stale cache should be a separate cache (for simplicity in using it, for example it would be great if systemd-resolve could flush the stale cache independently, and if SIGUSR1 dumped the 'real' and 'stale' caches separately). However I think @poettering is ok with the approach here. Anyway, I added a few specific review comments to the current commits.

Yeah, I think it's nice to maintain only a single cache, i.e. store everything only once. But always enforce the right rules when consuming the cache.

KiranVKumar commented:
> looks really good to me. Any chance you can add a test for this to TEST-75-RESOLVED?
> 
> also the commented default value in the .conf file still isnt right.

I couldn't find way to simulate timeout errors in the TEST-75-RESOLVED. 

> looks really good to me. Any chance you can add a test for this to TEST-75-RESOLVED?
> 
> also the commented default value in the .conf file still isnt right.

Since we are utilizing the Knot server in the TEST-75-RESOLVED, I haven't been able to find a method to simulate DNS server errors such as timeout, servfail, etc.

poettering commented:
> I couldn't find way to simulate timeout errors in the TEST-75-RESOLVED.

in the qemu case you could simply block UDP+TCP port 53 traffic via iptables. (it's ok if this part of the test is skipped for the nspawn test)

poettering commented on src/resolve/resolved-link.c:
```diff
@@ -726,7 +726,8 @@ DnsServer* link_set_dns_server(Link *l, DnsServer *s) {
 
-        if (l->unicast_scope)
+        /* Skip flushing the cache if server stale feature is enabled. */
+        if (l->unicast_scope && l->manager->stale_retention_usec == 0)
```
this should not be part of the test commit, of course

poettering replied:
hmm i think this simply fix will actually be fine. but could you document that in the man page. i.e. where the StaleRetentionSec= option is described, simply say "If this option is enabled, cache flushes when changing servers are disabled."

KiranVKumar replied:
updated man with "If this option is enabled, the cache will not be flushed when changing servers."

poettering commented:
shellcheck doesn't like the shell code.

but other than that actually looks fine to me

poettering commented on test/TEST-75-RESOLVED/test.sh:
```diff
@@ -36,6 +42,9 @@ test_append_files() {
     image_install delv dig host nslookup
+
+    # Install nftables
+    image_install nft
```
hmm, i just realized that this test actually sets TEST_NO_QEMU=1 and i am wondering why...

and what confuses me in particular is that you can do nft from inside a container. I had asumed this wouldn't work. /me is puzzled about this.

poettering replied:
so apparently nft just works in network namespaces, hence all good. (just checked with some namespacing kernel folks upstream kernel)

bluca commented on test/TEST-75-RESOLVED/test.sh:
```diff
@@ -17,6 +17,12 @@ if ! knotc -c "${TEST_BASE_DIR:?}/knot-data/knot.conf" conf-check; then
 fi
 
+# We need nftables to test the serve stale feature
+if ! nft -V; then
```
Rather than doing this, please check around the installation bit below, and then check again at runtime and skip the new test body only. That way we don't lose coverage entirely just because nft is missing/unusable.

bluca replied:
also, the way we check for these is `if command -v nft >/dev/null; then` as it might be present but not runnable for any number of reasons

bluca commented on test/units/testsuite-75.sh:
```diff
@@ -503,6 +503,7 @@ grep -qF "myservice.untrusted.test:1234" "$RUN_OUT"
 grep -qF "10.0.0.123" "$RUN_OUT"
 grep -qF "fd00:dead:beef:cafe::123" "$RUN_OUT"
 
+
```
spurious new line

poettering commented:
lgtm. thanks!

bluca commented:
test failed on s390x:

```
7014s testsuite-75.sh[823]: + dig stale1.unsigned.test -t A
7029s testsuite-75.sh[824]: ;; communications error to 127.0.0.53#53: timed out
7029s testsuite-75.sh[824]: ;; communications error to 127.0.0.53#53: timed out
7029s testsuite-75.sh[824]: ;; communications error to 127.0.0.53#53: timed out
7029s testsuite-75.sh[824]: ; <<>> DiG 9.18.12-0ubuntu0.22.04.1-Ubuntu <<>> stale1.unsigned.test -t A
7029s testsuite-75.sh[824]: ;; global options: +cmd
7029s testsuite-75.sh[824]: ;; no servers could be reached
```
